README.md
# I am looking for a contributor to this repository. I don't need this node anymore and didn't want to develop it any more. However, I can host it myself, so save yourself the inquiries whether you can host it!

# This component has some problems if the network connection to the PLC is interrupted, the PLC is not running, incorrect data is stored for the PLC or there is a deployment in the node red.
Everyone can take care of the problems, eliminate them and make a request.
But let that cry in the issues.

# node-red-contrib-ads

> A Node-RED implementation for the Twincat ADS protocol.
> (Twincat and ADS provided by Beckhoff &copy;. I'm not affiliated.)

Beckhoff TwinCat ADS support for Node-Red. Provides nodes to talk to TwinCat PLC variables over ADS.

Uses the NODE-ADS library for Node.JS (https://www.npmjs.com/package/node-ads).

It's tested with TwinCat 2 but its also works with TwinCat 3. Look at the Port.

For the latest updates see the [CHANGELOG.md](https://github.com/PLCHome/node-red-contrib-ads/blob/master/CHANGELOG.md)

## issue

If you have problems do not be afraid to open an issue.

If you open a new issue please remember that I do not know your hardware, software, PLC and programming.

Therefore it would be nice if you answer the following questions:

```
node-red-contrib-ads version:
nodered version:
on which operating system is the Node-RED running:
on which hardware system is the Nodered running: (PC/Raspberry)
PLC type:
Twincat version of the PLC:
Twincat version of the development environment:
```

## Installation

```sh
cd ~/.node-red
npm install node-red-contrib-ads
```

## Requirements
* Beckhoff PLC that has an ethernet connection and is connected to your LAN
    * Make your you give the PLC a fixed IP address
    * Make sure you can ping the PLC from another computer

## Configuration

1. Enable ADS on your PLC project. To do this click on your task and then enable the checkbox before `Create symbols` (if he is not disabled).
In addition, you can still, under I/O Devices click on Image and go to the ADS tab. Check the `Enable ADS Server` and also `Create symbols`.
Download the new configuration and make sure you reboot your PLC. The reboot is only needed when you are using TwinCat 2.

2. Now add a static route to our Beckhoff PLC. The route should point to your server that will run the proxy application.
It's also a good idea to add an extra static route that points to your local development device. This way you can test out the proxy from your development device too.

### Attention

1. TwinCAT AMS Router doesn't allow multiple TCP connections from the same host. So when you use two AdsLib instances on the same host to connect to the same TwinCAT router, you will see that TwinCAT will close the first TCP connection and only respond to the newest. If you start the TwinCat System Manager and Node-Red ADS on the same PC at the same time, Node-Red will not run anymore. You can set up a second IPv4 on the PC and assign to this a ADS NET ID under Twincat

2. As ADS is transmitted over a TCP connection, there is no real time guarantee.


## Global variables
Global variables must start with a dot: ```.engine```
```
VAR_GLOBAL
	engine		AT %QX0.0:	BOOL;
	deviceUp	AT %QX0.1:	BOOL;
	deviceDown	AT %QX0.2:	BOOL;
	timerUp:					TON;
	timerDown:				TON;
	steps:					BYTE;
	count:					UINT := 0;
	devSpeed:				TIME := t#10ms;
	devTimer:				TON;
	switch: BOOL;
END_VAR
```

Program variables must start with the programname: ```MAIN.UpTyp.timerUp.PT```



## Nodes added by this package

#### - ads-connection

A node that represents a TwinCat ADS device.
Be sure to create one for an PLC.

```
Host: IP address of the PLC
Client: IP address of the client if there is more than one. If it is empty, it will be determined automatically.
Target NetId: ADS NET ID of the PLC in the format 192.168.2.5.1.1
Source NetId: ADS NET ID for node red in the format 192.168.2.10.1.1 the same as the one added to the static route in the Beckhoff PLC.
Port: Normally 48898 for TwinCat 2/3
Target Port: Normally 801 for TwinCat 2 Runtime 1 or 851 for TwinCat 3 Runtime 1
Source Port: Normally 32905 for TwinCat 2/3
Timeout ms: The timeout in milliseconds, by default 500
```

#### - ADS System

Twincat ADS node that give you information about the PLC and PLC state.

##### input
You can write anything on the input, then the system state is pushed to the output.

##### output
```
msg.payload : Object
{ "connectState":2,
  "connectStateText":"CONNECTED",
  "adsState":5,
  "adsStateText":"RUN",
  "majorVersion":2,
  "minorVersion":11,
  "versionBuild":2605,
  "version":"2.11.2605",
  "deviceName":"TCatPlcCtrl",
  "symTab":138
}
```
###### Connect state (text)
The following connect state values are possible:
```
ERROR:                     -1
DISCONNECTED:               0
CONNECTING:                 1
CONNECTED:                  2
DISCONNECTING:              3
```

###### ADS state (text)
The following ADS state values are possible:
```
INVALID:      0
IDLE:         1
RESET:        2
INIT:         3
START:        4
RUN:          5
STOP:         6
SAVECFG:      7
LOADCFG:      8
POWERFAILURE: 9
POWERGOOD:    10
ERROR:        11
SHUTDOWN:     12
SUSPEND:      13
RESUME:       14
CONFIG:       15
RECONFIG:     16
STOPPING:     17
```

###### Version
This is the version of the PLC runtime. Either in single parts ("majorVersion", "minorVersion", "versionBuild") or assembled ("version")

###### Devicename
The Name of the PLC-Device

###### symTab
The PLC internal version number of the variable Händel assignment


#### - ADS Out

Twincat ADS output node that can send values to the PLC.

Enter the name of the variable, the type and the property name. If ads is connected the value is written to the PLC

If a topic is entered, the value is only sent if the topic is the same.

There is an possible to override the node configuration. You can add an property config to the message object. Not all configuration properties need to be overridden.
- `config.useIndex` : (boolen) override the variable use ID (true is not recommended for Twincat)
- `config.varName`: (string) override the variable name
- `config.indexGroup`: (interger/hex-string) override the variable indexGroup only used by useIndex varName is unused then
- `config.indexOffset`: (interger/hex-string) override the variable indexOffset only used by useIndex varName is unused then
- `config.varType`: (string) override the variable type
- `config.varSize`: (integer) the length on RAW and STRING type
- `config.timezone`: (string) only on date and time type 'TO_LOCAL' or 'UNCHANGED'
- `config.outProperty`: (string) the property for the outvalue
- `config.topic`: (string) the topic being checked against the message topic. It can be '' to delete an topic.


#### - ADS In

Twincat ADS input node that can recive values from the PLC.

This note is not for the transmission of cyclic data. Do not use it in conjunction with a timer. There is a notification node for this.

Enter the name of the variable, the type and the property name for the output.
You can still decide whether a new output with the property for the value will be created or the property will be inserted into the inputvalue and output at the output.

There is an possible to override the node configuration. You can add an property config to the message object. Not all configuration properties need to be overridden.
- `config.useIndex` : (boolen) override the variable use ID (true is not recommended for Twincat)
- `config.varName`: (string) override the variable name
- `config.indexGroup`: (integer/hex-string) override the variable indexGroup only used by useIndex varName is unused then
- `config.indexOffset`: (integer/hex-string) override the variable indexOffset only used by useIndex varName is unused then
- `config.varType`: (string) override the variable type
- `config.varSize`: (integer) the length on RAW and STRING type
- `config.isarray`: (bool) is an ARRAY type, varLowIndex and varHighIndex must set
- `config.varLowIndex`: (integer) the Low index on ARRAY type
- `config.varHighIndex`: (integer) the High index on ARRAY type
- `config.timezone`: (string) only on date and time type 'TO_LOCAL' or 'UNCHANGED'
- `config.inProperty`: (string) the property for the outvalue
- `config.useInputMsg`: (bool) the input message will be used to build the out message or an new message will be build.
- `config.topic`: (string) the topic being checked against the message topic. It can be '' to delete an topic.

#### - ADS Notification

Twincat ADS input node that can automatically recive values from the PLC, if they change.
Beckhoff says you should not have more than 510 variables monitored.
Each time the ADS-Node is connected to the PLC, the PLC automatically sends the value once.

Note, if you have multiple Notification with the same variable name then only one of the nodes will be registered in the PLC. Which is coincidence. therefore, the same variable may not be "cyclic" and "onchange" at the same time. Or number and bool at the same time. Please use the link node from the node red base nodes.

- Transmission Mode: With "cyclic", the variable is polled cyclically by the PLC and transmitted, with "onchange" the PLC monitors.
- Max Delay:  At the latest after this time, the ADS Device Notification is called. The unit is 1ms.
- Cycle Time:  The ADS server checks if the value changes in this time slice. The unit is 1ms


#### - ADS Symbols

This note loads a list of all symbols or types from the PLC when written to the input. If a string or string array is written, the output is filtered by name. Wildcards [?=char;*=chars] are possible.

This example makes an web ui for inspecting the PLC. The ADS Connection still needs to be set.
```json

CHANGELOG.md
### 1.1.30: Maintenance Release

**Fixes**
- Improvement for check ads netID

### 1.1.29: Maintenance Release

**Fixes**
- Improvement for connect, disconnect and deploy

### 1.1.28: Maintenance Release

**Enhancements**
- read/write/notification: ADS Read to indexOffset and indexGroup #33

### 1.1.27: Maintenance Release

**Fixes**
- Target & source port swapped in read.me #37
- API (Twincat) not longer in the connection class, now it is global
- Improves reconnect and lost network

### 1.1.26: Maintenance Release

**Fixes**
- some string.length behavior in v14.13.1 symbols and types crashes

### 1.1.25: Maintenance Release

**Fixes**
- some trouble on deploy

### 1.1.24: Maintenance Release

**Enhancements**
- add read arrays on in node and notification node
- enter the client's IP address if there is more than one

**Fixes**
- some trouble on data error and end the socket
- some trouble on restart after data error

### 1.1.23: Maintenance Release

**Enhancements**

- something about issue
- read/write/notification: on error debug points to the correct node and varame in debug message

### 1.1.22: Maintenance Release

**Fixes**

- String output: The "size" argument must be of type number. Received type string #19

### 1.1.21: Maintenance Release

**Fixes**

- ADS notification not working after PLC has been offline (same as #14) #18

### 1.1.20: Maintenance Release

**Enhancements**

- ads-connection: adjustable timeout

**Fixes**

- node-ads-api: version 1.4.13 used
- ads-out: check for undefined

### 1.1.19: Maintenance Release

**Fixes**

- node-ads-api: version 1.4.11 used

### 1.1.18: Maintenance Release

**Fixes**

- node-ads-api: version 1.4.10 used

### 1.1.17: Maintenance Release

**Fixes**

- node-ads-api: version 1.4.9 used


### 1.1.16: Maintenance Release

**Fixes**

- ads out, ads in: override intern config with msg.config: corrected varType and timezone


### 1.1.15: Maintenance Release

**Enhancements**

- issue #11 multiple notifications with the same variable
- ads out, ads in: override intern config with msg.config

**Fixes**

- Release ADS-NOTIFICATION on deploy modified Node / modified Flows
- PLC relieved, live sign increased to 10 sec.


### 1.1.14: Maintenance Release

**Enhancements**

- Cache datatyps and symbols
- Topic added for all nodes.
- If a topic is entered at ADS-OUT, the value is only sent if the topic is the same.


**Fixes**

- Exception when use Time or Date on ADS-NOTIFICATION


### 1.1.13: Maintenance Release

**Enhancements**

- debug added
```
  debug=node-red-contrib-ads:*
  debug=node-red-contrib-ads:*,-node-red-contrib-ads:adsConnectionNode:Cyclic
  debug=node-red-contrib-ads:adsConnectionNode
  debug=node-red-contrib-ads:adsConnectionNode:Cyclic
  debug=node-red-contrib-ads:adsNotificationNode
  debug=node-red-contrib-ads:adsInNode
  debug=node-red-contrib-ads:adsOutNode
  debug=node-red-contrib-ads:adsSystemNode
  debug=node-red-contrib-ads:adsSymbolsNode
```


**Fixes**

- ADS ads-symbols on TC3
- [Node.js under heavy load when connection down #6](https://github.com/PLCHome/node-red-contrib-ads/issues/6)
- node-ads-api 1.4.7: Notification result may be empty when the connection is closed


### 1.1.12: Maintenance Release

**Enhancements**

- Example for web ui with Node for get symbols and types from ADS.


**Fixes**

- ADS Notification Note wont work


### 1.1.11: Maintenance Release

**Enhancements**

- Node for get symbols and types from ADS
- Nodes show Varname instead of Node-name
- revision README.md
- Use input msg for output in ADS In


**Fixes**

- README.md wrong link CHANGELOG.md


### 1.1.10: Maintenance Release

**Enhancements**

- Revision README.md
- Message property for ADS-OUT/ADS-NOTIFICATION


**Fixes**

- Node red does not start if the port number is too large
- System-state on timer 50 ms decoupled so that it sends less frequently
- Set system-state to INVALID, when connect-state not RUN
- Exception by get a message property on ADS-IN


ads-connection.js
'use strict'
var nodeads = require('node-ads-api')
var adsHelpers = require('./ads-helpers')
var util = require('util')
var debug = require('debug')('node-red-contrib-ads:adsConnectionNode')
var debugCyclic = require('debug')('node-red-contrib-ads:adsConnectionNode:Cyclic')
var twincat={}

module.exports = function (RED) {

  function adsConnectionNode(config) {
    RED.nodes.createNode(this, config)
    var node = this
    debug('config:',config)

    node.system = {}
    internalSetConnectState(adsHelpers.connectState.DISCONNECTED)
    internalSetAdsState(nodeads.ADSSTATE.INVALID)

    node.notificationNodes = []
    node.notificationSubscribed = {}
    node.systemNodes = []

    function removeClient(){
      debug('removeClient:','Client:',!(!twincat.adsClient),'symbolsCache:',!(!twincat.symbolsCache),'datatypsCache:',!(!twincat.datatypsCache))
      node.notificationSubscribed = {}
      delete(twincat.adsClient)
      delete(twincat.symbolsCache)
      delete(twincat.datatypsCache)
    }
    
    /* connect to PLC */
    function connect() {
      internalSetConnectState(adsHelpers.connectState.CONNECTING)
      startTimer(45000)
      var adsoptions = {}

      adsoptions.host = config.host
      if (config.clientIP !== "") {
        adsoptions.localAddress = config.clientIP
      }
      adsoptions.amsNetIdTarget = config.amsNetIdTarget
      adsoptions.amsNetIdSource = config.amsNetIdSource
      adsoptions.port = parseInt(config.port)
      adsHelpers.checkPort(node,adsoptions.port,48898)
      adsoptions.amsPortSource = parseInt(config.amsPortSource)
      adsHelpers.checkPort(node,adsoptions.amsPortSource,801)
      adsoptions.amsPortTarget = parseInt(config.amsPortTarget)
      adsHelpers.checkPort(node,adsoptions.amsPortTarget,32905)
      adsoptions.timeout = parseInt(config.adsTimeout||500)
      if (adsoptions.timeout < 500) adsoptions.timeout = 500
      debug('connect:',adsoptions)
      if (twincat.adsClient) {
        startTimer(5000)
      } else {
        twincat.adsClient = nodeads.connect(adsoptions,
          function (){
            twincat.adsClient.readDeviceInfo(function (err,handel) {
              if (err) {
                removeClient()
                node.error('Error on connect: check target NetId or routing')
                internalSetConnectState(adsHelpers.connectState.ERROR)
                startTimer(120000)
                debug('connect:','readDeviceInfo:',err)
              } else {
                debug('connect:','readDeviceInfo:',handel)
                node.system.majorVersion = handel.majorVersion
                node.system.minorVersion = handel.minorVersion
                node.system.versionBuild = handel.versionBuild
                node.system.version = handel.majorVersion+'.'+handel.minorVersion+'.'+handel.versionBuild
                node.system.deviceName = handel.deviceName
                internalSystemUpdate()
                internalSubscribeLiveTick()
                startTimer(120000)
                internalSubscribeSymTab()
                function subscribe(i,cb){
                  if (i<node.notificationNodes.length) {
                    internalSubscribe(node.notificationNodes[i],()=>{subscribe(++i,cb)})
                  } else cb()
                }
                subscribe(0,()=>{internalSetConnectState(adsHelpers.connectState.CONNECTED)})
                //node.notificationNodes.forEach(internalSubscribe)
                internalSetConnectState(adsHelpers.connectState.CONNECTED)
              }
            })
          }
        )
      }

      twincat.adsClient.on('notification', function (handle){
          if (handle.indexGroup == nodeads.ADSIGRP.DEVICE_DATA &&
              handle.indexOffset == nodeads.ADSIOFFS_DEVDATA.ADSSTATE) {
            debugCyclic('notification:','connectCheckADSSTATE:', handle.indexGroup, handle.indexGroup, 
                           'handle.indexOffset', handle.indexOffset, 'handle.value', handle.value)
            if (node.system.connectState == adsHelpers.connectState.CONNECTED) {     
              startTimer(10000)
            }
            internalSetAdsState(handle.value)
          } else if (handle.indexGroup == nodeads.ADSIGRP.SYM_VERSION &&
                     handle.indexOffset == 0) {
            debug('notification:','CheckSYM_VERSION:', 'handle.indexGroup:', handle.indexGroup, 
                           'handle.indexOffset', handle.indexOffset, 'handle.value', handle.value)
            if (node.system.symTab != handle.value) {
              node.system.symTab = handle.value
              internalSystemUpdate()
            }
            if (handle.start) {
              handle.start = false
            } else {
              node.log(util.format('new sym Version - restart'))
              internalRestart(connect)
            }
          }
          if (node.notificationSubscribed[handle.symname]) {
            debug('notification:',handle.symname, 'handle.notifyHandle', handle.notifyHandle,
                         'handle.totalByteLength', handle.totalByteLength, 'handle.symhandle', handle.symhandle, 'handle.value', handle.value, 
                         'handle.bytelength', handle.bytelength)
            node.notificationSubscribed[handle.symname].map((n)=> {
              if (n.onAdsData) {
                n.onAdsData(handle)
              }
            })
          }
        }
      )


      twincat.adsClient.on('error', function (error) {
          debug('onerror:',error)
          if (error){
            node.error(util.format('Error ADS: %s', error))
            if (!twincat.connected) {
              if (node.system.connectState == adsHelpers.connectState.CONNECTING) {
                internalSetConnectState(adsHelpers.connectState.ERROR)
              }
              internalRestart(() => {startTimer(60000)})
            }
          }
        }
      )

    }
    /* end connect to PLC */

    /* check connection to PLC */
    var conncetTimer

    function startTimer(time){
      const defTime = 60000
      debugCyclic('startTimer:','clearTimeout')
      clearTimeout(conncetTimer)
      if (node.system.connectState != adsHelpers.connectState.DISCONNECTING) {
        debugCyclic('startTimer:',time||defTime)
        conncetTimer = setTimeout(function () {
          debug('startTimer:','Timeout')
          if (node.system.connectState == adsHelpers.connectState.CONNECTED) {
            internalSetConnectState(adsHelpers.connectState.DISCONNECTED)
          }
          internalRestart(connect)
        },time||defTime)
      }
    }
    /* needs internalSubscribeLiveTick */
    /* end check connection to PLC */

    /* for ads-notification */
    node.subscribe = function (n) {
      debug('subscribe:',n.id, n.type)
      if (node.notificationNodes.indexOf(n) < 0) {
        node.notificationNodes.push(n)
      }
      if (node.system.connectState == adsHelpers.connectState.CONNECTED) {
        internalSubscribe(n)
      }
    }

    node.unsubscribe = function (n,cb) {
      debug('unsubscribe:',n.id, n.type, n.notifyHandle)
      var index = node.notificationNodes.indexOf(n)
      if (index >= 0) {
        node.notificationNodes.splice(index,1)
      }
      if ((node.notificationSubscribed[n.symname])){
        var index = node.notificationSubscribed[n.symname].indexOf(n)
        if (index >= 0) {
          node.notificationSubscribed[n.symname].splice(index,1)
        }
        if (!node.notificationSubscribed[n.symname].length){
          delete node.notificationSubscribed[n.symname]
        }
      }
      if (n.notifyHandle !== undefined && (!node.notificationSubscribed[n.symname])) {
        var handle =  {
          notifyHandle: n.notifyHandle
        }
        if (twincat.adsClient) {
          twincat.adsClient.releaseNotificationHandle(handle, function() {
            delete(n.notifyHandle)
            debug('unsubscribe: done')
            if (cb){
              cb()
            }
          })
        } else {
          debug('unsubscribe: twincat.adsClient deleted')
          if (cb){
            cb()
          }
        }
      } else {
        debug('unsubscribe: cat release handle:',notifyHandle)
        if (cb){
          cb()
        }
      }
  }
    
    /* end for ads-notification */

    /* subscribe on PLC */
    function internalSubscribe(n,cb){
      debug('internalSubscribe:',n.id, n.type)
      var handle =  {
        transmissionMode: nodeads.NOTIFY[n.transmissionMode],
        maxDelay: n.maxDelay,
        cycleTime: n.cycleTime,
      }
      if (n.useIndex) {
        handle.indexGroup = n.indexGroup
        handle.indexOffset = n.indexOffset
      } else {
        handle.symname = n.symname
      }
      if (adsHelpers.isRawType(n.adstype)) {
        handle.bytelength = parseInt(n.bytelength)
      } else {
        if (n.array) {
          if (adsHelpers.isStringType(n.adstype) && n.bytelength) {
            handle.bytelength = nodeads.array(nodeads.string(parseInt(n.bytelength)),parseInt(n.lowindex),parseInt(n.highindex))
          } else {
            handle.bytelength = nodeads.array(nodeads[n.adstype],parseInt(n.lowindex),parseInt(n.highindex))
          } 
        } else {
          if (adsHelpers.isStringType(n.adstype) && n.bytelength) {
            handle.bytelength = nodeads.string(parseInt(n.bytelength))
          } else {
            handle.bytelength = nodeads[n.adstype]
          }
        }
      }
      if (adsHelpers.isTimezoneType(n.adstype)) {
        handle.useLocalTimezone = (n.timezone === "TO_LOCAL")
      }
      if (!(node.notificationSubscribed[n.symname])){
        if (twincat.adsClient) {
          debug('internalSubscribe:',handle)
          node.notificationSubscribed[n.symname]=[]
          node.notificationSubscribed[n.symname].push(n)
          twincat.adsClient.notify(handle, function(err){
            if (err){
              n.error(util.format("Ads Register Notification '%s' %s",n.symname, err))
            } else {
              node.notificationSubscribed[n.symname].map((no)=>{
                no.notifyHandle = handle.notifyHandle
              })
            }
            if (cb){
              cb()
            }
          })
        } else if (cb) cb()
      } else {
        var index = node.notificationSubscribed[n.symname].push(n)
        debug('internalSubscribe: reuse notifyHandle',node.notificationSubscribed[n.symname][0].notifyHandle)
        if (node.notificationSubscribed[n.symname][0].notifyHandle) {
          node.notificationSubscribed[n.symname].map((n)=>{
            if (!n.notifyHandle) {
              n.notifyHandle = node.notificationSubscribed[n.symname][0].notifyHandle
            }
          })
        }
        if (cb) cb()
      }
    }

    function internalSubscribeLiveTick(){
      debug('internalSubscribeLiveTick:')
      var handle =  {
        indexGroup: nodeads.ADSIGRP.DEVICE_DATA,
        indexOffset: nodeads.ADSIOFFS_DEVDATA.ADSSTATE,
        transmissionMode: nodeads.NOTIFY.CYCLIC,
        cycleTime: 1000,
        bytelength: nodeads.WORD
      }
      if (twincat.adsClient) {
        debug('internalSubscribeLiveTick:',handle)
        twincat.adsClient.notify(handle, function(err){
          if (err){
            node.error(util.format('Ads Register Notification live tick %s', err))
          }
        })
      }
    }

    function internalSubscribeSymTab(){
      debug('internalSubscribeLiveTick:')
      var handle =  {
        indexGroup: nodeads.ADSIGRP.SYM_VERSION,
        indexOffset: 0,
        transmissionMode: nodeads.NOTIFY.ONCHANGE,
        bytelength: nodeads.BYTE,
        start: true
      }

      if (twincat.adsClient) {
        debug('internalSubscribeSymTab:',handle)
        twincat.adsClient.notify(handle, function(err){
          if (err){
            node.error(util.format('Ads Register Notification Sym Tab %s', err))
          }
        })
      }
    }
    /* end subscribe on PLC */

    /* write to PLC */
    node.write = function (n, config, value) {
      debug('write:',n.id, n.type)
      if (node.system.connectState == adsHelpers.connectState.CONNECTED) {
        var handle =  {
          propname: 'value',
          value: value
        }
        if (config.useIndex) {
          handle.indexGroup = config.indexGroup
          handle.indexOffset = config.indexOffset
        } else {
          handle.symname = config.symname
        }
        if (adsHelpers.isRawType(config.adstype)) {
          handle.bytelength = parseInt(config.bytelength)
        } else {
          if (adsHelpers.isStringType(config.adstype) && config.bytelength) {
            handle.bytelength = nodeads.string(parseInt(config.bytelength))
          } else {
            handle.bytelength = nodeads[config.adstype]
          }
        }
        if (adsHelpers.isTimezoneType(config.adstype)) {
          handle.useLocalTimezone = (config.timezone === "TO_LOCAL")
        }
        if (twincat.adsClient) {
          debug('write:',handle)
          twincat.adsClient.write(handle,
            function (err){
              if (err) {
                n.error(util.format("Ads write '%s' %s", config.symname,err))
              }
            } )
        }
      }
    }
    /* end write to PLC */

    /* read from PLC */
    node.read = function (n, config, cb) {
      debug('read:',n.id, n.type)
      if (node.system.connectState == adsHelpers.connectState.CONNECTED) {
        var handle =  {
          propname: 'value'
        }
        if (config.useIndex) {
          handle.indexGroup = config.indexGroup
          handle.indexOffset = config.indexOffset
        } else {
          handle.symname = config.symname
        }
        if (adsHelpers.isRawType(config.adstype)) {
          handle.bytelength = parseInt(config.bytelength)
        } else {
          if (config.array) {
            if (adsHelpers.isStringType(config.adstype) && config.bytelength) {
              handle.bytelength = nodeads.array(nodeads.string(parseInt(config.bytelength)),parseInt(config.lowindex),parseInt(config.highindex))
            } else {
              handle.bytelength = nodeads.array(nodeads[config.adstype],parseInt(config.lowindex),parseInt(config.highindex))
            } 
          } else {
            if (adsHelpers.isStringType(config.adstype) && config.bytelength) {
              handle.bytelength = nodeads.string(parseInt(config.bytelength))
            } else {
              handle.bytelength = nodeads[config.adstype]
            }
          }
        }
        if (adsHelpers.isTimezoneType(config.adstype)) {
          handle.useLocalTimezone = (config.timezone === "TO_LOCAL")
        }
        if (twincat.adsClient) {
          debug('read:',handle)
          twincat.adsClient.read(handle, function(err, handle){
            if (err) {
              n.error(util.format("Ads read '%s' %s",config.symname, err))
            } else {
              cb(handle)
            }
          })
        }
      }
    }
    /* end read from PLC */

    /* node RIP */
    node.on('close', function (done) {
      debug('close:')
      internalRestart(done)
    })

    function internalRestart(done) {
      debug('internalRestart:','enter')
      internalSetConnectState(adsHelpers.connectState.DISCONNECTING)
      clearTimeout(conncetTimer)
      if (twincat.adsClient) {
        let ads = twincat.adsClient
        delete(twincat.adsClient)
        ads.end(function (){
          internalSetConnectState(adsHelpers.connectState.DISCONNECTED)
          removeClient()
          var sleep = setTimeout(function () {
            clearTimeout(sleep)
            debug('internalRestart:','done')
            done()
          },1000)
        })
      } else {
        debug('internalRestart:','done no adsClient')
        internalSetConnectState(adsHelpers.connectState.DISCONNECTED)
        done()
      }
    }
    /* end node RIP */

    /* set Note State */
    function internalSetConnectState (cState) {
      debug('internalSetConnectState:','new state:'+cState,'old state:'+node.system.connectState)
      if ((!node.system.connectState) || node.system.connectState != cState) {
        debug('internalSetConnectState:','set state:'+cState)
        node.system.connectState = cState
        node.system.connectStateText = adsHelpers.connectState.fromId(cState)
        if (node.system.connectState != adsHelpers.connectState.CONNECTED) {
          internalSetAdsState(nodeads.ADSSTATE.INVALID)
        }
        internalSystemUpdate()
      }
    }
    /* end Note State */

    /* symbols from PLC */
    node.getSymbols = function (force,cb) {
      debug('getSymbols:','enter')
      if (!force && (twincat.symbolsCache)) {
        debug('getSymbols by cache')
        cb(twincat.symbolsCache)
      } else {
        if (node.system.connectState == adsHelpers.connectState.CONNECTED) {
          if (twincat.adsClient) {
            twincat.adsClient.getSymbols( function (err, symbols){
              debug('getSymbols:',err,symbols)
              if (err) {
                n.error(util.format('Ads Symbols %s', err))
              } else {
                twincat.symbolsCache = symbols
                cb(symbols)
              }
            }, true )
          }
        }
      }
    }

    node.getDatatyps = function (force,cb) {
      debug('getDatatyps:','enter')
      if (!force && (twincat.datatypsCache)) {
        debug('getDatatyps by cache')
        cb(twincat.datatypsCache)
      } else {
        if (node.system.connectState == adsHelpers.connectState.CONNECTED) {
          if (twincat.adsClient) {
            twincat.adsClient.getDatatyps( function (err, datatyps){
              debug('getDatatyps:',err,datatyps)
              if (err) {
                n.error(util.format('Ads Datatyps %s', err))
              } else {
                twincat.datatypsCache = datatyps
                cb(datatyps)
              }
            }, true )
          }
        }
      }
    }
    /* end symbols from PLC */

    /* system Nodes */
    node.systemRegister = function (n){
      debug('systemRegister:',n.id,n.type)
      if (node.systemNodes.indexOf(n) < 0) {
        node.systemNodes.push(n)
        node.systemUpdate(n)
        setSystemStatus(n)
      }
    }

    node.systemUnregister = function (n){
      debug('systemUnregister:',n.id,n.type)
      var index = node.systemNodes.indexOf(n)
      if (index >= 0) {
        node.systemNodes.splice(index,1)
      }
    }
    
    function internalSystemUpdate () {
      debug('internalSystemUpdate:','enter')
      if (!(node.timerSU)) {
        node.timerSU = setInterval(function () {
          clearTimeout(node.timerSU)
          delete(node.timerSU)
          if (node.systemNodes) {
            node.systemNodes.forEach(function(n){
                debug('internalSystemUpdate:','call',n.id,n.type)
                node.systemUpdate(n)
                setSystemStatus(n)
              })
          }
        },50)
      }
    }

    node.systemUpdate = function (n) {
      if (n) {
        debug('systemUpdate:','enter',n.id,n.type)
        n.onData(node.system)
      }
    }

    function setSystemStatus (n) {
      debug('setSystemStatus:','enter')
      var fillSystem = "grey"
      var shapeSystem = "ring"
      var textSystem = node.system.connectStateText
      switch (node.system.connectState) {
        case adsHelpers.connectState.ERROR:
          fillSystem = "red"
          break
        case adsHelpers.connectState.DISCONNECTED:
          fillSystem = "grey"
          break
        case adsHelpers.connectState.CONNECTING:
        case adsHelpers.connectState.DISCONNECTING:
          fillSystem = "yellow"
          break
        case adsHelpers.connectState.CONNECTED:
          fillSystem = "green"
          shapeSystem = "dot"
          textSystem = node.system.adsStateText
          switch (node.system.adsState) {
            case nodeads.ADSSTATE.INVALID:
                  fillSystem = "grey"
                  break
            case nodeads.ADSSTATE.IDLE:
            case nodeads.ADSSTATE.RESET:
            case nodeads.ADSSTATE.INIT:
            case nodeads.ADSSTATE.POWERGOOD:
            case nodeads.ADSSTATE.SHUTDOWN:
            case nodeads.ADSSTATE.SUSPEND:
            case nodeads.ADSSTATE.RESUME:
            case nodeads.ADSSTATE.RECONFIG:
                  fillSystem = "blue"
                  break
            case nodeads.ADSSTATE.CONFIG:
            case nodeads.ADSSTATE.START:
            case nodeads.ADSSTATE.STOPPING:
            case nodeads.ADSSTATE.SAVECFG:
            case nodeads.ADSSTATE.LOADCFG:
                  fillSystem = "yellow"
                  break
            case nodeads.ADSSTATE.RUN:
                  fillSystem = "green"
                  break
            case nodeads.ADSSTATE.ERROR:
            case nodeads.ADSSTATE.STOP:
            case nodeads.ADSSTATE.POWERFAILURE:
                  fillSystem = "red"
                  break
          }
          break
      }
      debug('setSystemStatus:',n.id,n.type,{fill:fillSystem,shape:shapeSystem,text:textSystem})
      n.status({fill:fillSystem,shape:shapeSystem,text:textSystem})
    }

    function internalSetAdsState(adsState) {
      debugCyclic('internalSetAdsState:','new State:',adsState,'old State:',node.system.adsState)
      if ((!node.system.adsState) || node.system.adsState != adsState) {
        debug('internalSetAdsState:','set State:',adsState)
        node.system.adsState = adsState
        node.system.adsStateText = nodeads.ADSSTATE.fromId(adsState)
        internalSystemUpdate()
      }

    }
    /* end system Nodes*/
    connect()
  }
  RED.nodes.registerType('ads-connection', adsConnectionNode)

}
